### ISSUE TITLE
Make MethodTypes first-class

### ISSUE THREAD

**Issue by Blaisorblade** [lampepfl/dotty#4670]
Recently in #2500, @milessabin @LukaJCB and @LPTK advocated turning method types (with the existing syntax shown into method types) into value types. From my summary:

> With MethodTypes, [X]F[X] becomes internally Function0[F[_]] and then after erasure Function0 (and this is hidden from the typechecker). That is, method types are just functions, like in Haskell. This is just the standard and well-understood universal quantification in System F, and also should work fine with effects. It'd still be good to formalize this carefully.

Since such MethodTypes always need to be applied at runtime, they should avoid issues with polymorphic references that appeared in #2500.
This variant appears to avoid lots of other issues discussed there, and it seems what @odersky considered in https://github.com/lampepfl/dotty/pull/1560#issuecomment-252549931. He still said this would require a SIP tho, so that should be part of this issue.

**Comment by sstucki** [comment]
Quoting @LPTK's question from #2500, as I am wondering about the same thing...
 
>> With MethodTypes, `[X]F[X]` becomes internally `Function0[F[_]]`
>
> Isn't the latter equivalent to `Function0[F[X] forSome { type X }]`, so very different from the former?

**Comment by LukaJCB** [comment]
I'm confused, if `val foo: [X]F[X]` is equivalent to `def foo[X]: F[X]`, then it should be a universal quantifier, not an existential one, no?

**Comment by Blaisorblade** [comment]
I probably should just write that MethodTypes just always erase to FunctionTypes, even if they take no value arguments; that's less confusing.

I only wrote that (informally) to represent the first erasure step — I want to say that `[X]F[X]` erases to `Function0[F]` and then to `Function0`, but `Function0[F]` isn't a real thing and erasure doesn't happen step-by-step.

**Comment by fommil** [comment]
Not to state the obvious, but this will have implications for tooling that expects the old form (well, whatever tools that can be migrated to scala 3 that is).

**Comment by LPTK** [comment]
One reason I was [showing](https://github.com/lampepfl/dotty/issues/2500#issuecomment-397300984) there is a pretty direct encoding of polymorphic methods using implicit dependent function types is that I think they should behave pretty much the same, at least for users.
I.e., it should be possible to omit type abstraction when defining a method value, just like one can omit the implicit parameters of a lambda when they can be inferred from the expected type:

```scala
// Example:
def m[A](x: A): List[A] = a :: Nil
// made first-class:
val m: [A](A) => List[A] = a => a :: Nil
// or:                   = _ :: Nil
// or, explicitly:       = [A](a) => a :: Nil

// Similar to how we have:
def m(implicit x: T): List[T] = implicitly[T] :: Nil
// made first-class:
val m: implicit (x: T) => List[T] = implicitly[T] :: Nil
// or, explicitly:                = implicit (a: T) => implicitly[T] :: Nil
```

The encoding also tells us that insofar as dependent function types are sound, so should first-class methods be. A bit more generally:

```scala
def m[A >: AL <: AU, ...](x: F[A], ...): G[A] = ...
// made first-class:
val m: [A >: AL <: AU, ...](x: F[A], ...) => G[A] = (x, ...) => ...
// encoded:
val m: (A: { type T >: AL <: AU }, ...) => (x: A.T, ...) => G[A.T] = (A, ...) => (x, ...) => ...
```

This leads me to wonder if we couldn't make the analogy between type parameters and implicits even tighter. I'd go as far as to allow _value_ parameters to be made implicit the same way as for types, by enclosing them in square brackets, and propose that `[a:A]` be equivalent to `implicit(a:A)`.

We'd distinguish type parameters from value parameters by the case of the first letter (there is precedent for distinguishing meaning based on case in Scala and other languages).

This way, we can rewrite the implicit method example above as:

```scala
def m[x: T]: List[T] = implicitly[T] :: Nil
// made first-class:
val m: [x: T] => List[T] = implicitly[T] :: Nil
// or:                   = [a: T] => implicitly[T] :: Nil
```

which also provides a solution to the problems in https://github.com/lampepfl/dotty/issues/1260:

```scala
def f(implicit x: C): A => B = ...
// now expressed as:
def f[x: C]: A => B = ...
f(new A)    // okay
f[c](new A) // okay
```

**Comment by milessabin** [comment]
I think it's very helpful to think of implicit resolution as value inference, analogously to type inference.

**Comment by sstucki** [comment]
> I think it's very helpful to think of implicit resolution as value inference, analogously to type inference.

It's a tempting analogy, but also a potentially misleading one. As you know, type inference and implicit resolution employ quite different strategies (unification vs. proof search). So using similar syntax for the two features is likely to cause confusion, IMO.

(There are of course languages, such as Agda, that offer both "value inference" and "implicit resolution" [as separate mechanisms](https://doi.org/10.1145/2034773.2034796).)

**Comment by Blaisorblade** [comment]
@LPTK The analogy was tempting enough for me to propose `[[...]]` in 2013 (http://blaisorbladeprog.blogspot.com/2013/01/flexible-implicits-from-agda-for-scala.html), but many people seem to care about having only types in brackets. I'd still love to pick such a simple syntax for #1260 instead of inventing something new, we're just almost out of ASCII chars.

@fommil Obviously, that's a valid concern for all Dotty changes, but I guess you're seeing more than that?

**Comment by fommil** [comment]
@Blaisorblade yeah, I mean this is the sort of thing that IDEs might be grepping for in order to special case. But I guess everything is getting rewritten anyways. The other tools that come to mind are ammonite and scalameta.

**Comment by smarter** [comment]
I haven't followed the discussions here and in https://github.com/lampepfl/dotty/issues/2500, but today I did something you might find interesting: https://github.com/lampepfl/dotty/pull/4672

**Comment by LPTK** [comment]
@smarter this looks really promising!

Here are two more use cases for the feature:

**1. Make polymorphic case classes extend the appropriate function types:**

This used to be a puzzler for me when I started:

```scala
case class A(x: Int)
case class B[T](x: Int)
// ^ A implicitly extends Function1, but B does not for lack of polymorphic function type!
val a: Int => A = A       // okay
val b: Int => B[Int] = B  // nope, found: B.type; required: Int => B[Int]
```

It's particularly annoying when using a pipe operator `|>` as in `x |> Some.apply` because the shorter `x |> Some` currently won't work (though `x |> A` works).


**2. Make partial application much more flexible:**

I saw you mention polymorphic eta-expansion, so we could write (given class `B` above) `val b = B` of inferred type `[T] -> Int => B[T]`.

But do you think this could be generalized so that when a type argument cannot be conclusively inferred, instead of picking `Nothing` or `Any`, just make the result polymorphic!

```scala
val f = List(1,2,3).map _
// f: [B] -> (Int => B) => List[B]

val g = Some(1).fold(Nil) _
// g: [B >: Nil.type] -> (Int => B) => B
```

It would be nice being able to omit the `_` above, but I imagine it could break a lot of code and make type-checking slower as we would start inferring many more polymorphic types everywhere.

**Comment by smarter** [comment]
In Dotty we're trying to make `_` for eta-expansion obsolete, `val x = f` should behave like `val x = f _`. I don't know if we want eta-expansion without an expected type to be polymorphic or not, like you said we'd have to check how much code that breaks.

**Comment by odersky** [comment]
The SI calculus in "Simplicitly" (POPL 2018) makes eta expansion monomorphic, since its restricted (R) types are monomorphic. This is required since otherwise it would be ambiguous where generalization and instantiation happens.
